### PR TITLE
poc migration

### PR DESCRIPTION
**Pull request by chenqianfzh**
This PR migrates the PoC work presented in last release, which is saved in branch https://github.com/CentaurusInfra/fornax/tree/backup-main.

Explanation about this code change is presented in https://github.com/pdgetrf/ArktosEdge/blob/main/slides/Fornax%20code%20change.pptx.

### CODE REVIEW

File: cloud/pkg/edgecontroller/controller/downstream.go [hunk 1]
```diff
@@ -92,7 +104,7 @@ func (dc *DownstreamController) syncPod() {
 			if err := dc.messageLayer.Send(*msg); err != nil {
 				klog.Warningf("send message failed with error: %s, operation: %s, resource: %s", err, msg.GetOperation(), msg.GetResource())
 			} else {
-				klog.V(4).Infof("send message successfully, operation: %s, resource: %s", msg.GetOperation(), msg.GetResource())
+				klog.V(4).Infof("message sent successfully, operation: %s, resource: %s", msg.GetOperation(), msg.GetResource())
```
Comment (pdgetrf) [inline cloud/pkg/edgecontroller/controller/downstream.go:107]: "message sent successfully, operation: %s, resource: %s" this is repeated multiple times. consider putting it into a common string var.

File: cloud/pkg/edgecontroller/controller/upstream.go [hunk 2]
```diff
@@ -546,6 +560,243 @@ func (uc *UpstreamController) updateNodeStatus() {
 	}
 }
 
+// createEdgeCluster create a new edgeCluster
```
Comment (pdgetrf) [inline cloud/pkg/edgecontroller/controller/upstream.go:563]: the function name is pretty good already. this comment doesn't seem needed.

File: cloud/pkg/missionstatepruner/pruner.go [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+			default:
+				// remove the mission state about the sub-clusters of the dead edgecluster
+				if deadEdgeClusters[parts[0]] {
+					delete(mission.State, key)
+					changed = true
+				}
+			}
+		}
+		if changed {
+			klog.V(3).Infof("Pruning the state of mission %v ...", mission.Name)
```
Comment (pdgetrf) [inline cloud/pkg/missionstatepruner/pruner.go:124]: what does "Pruning" mean here?

File: cloud/pkg/synccontroller/objectsync.go [hunk 4]
```diff
@@ -71,6 +93,12 @@ func (sctl *SyncController) manageObject(sync *v1alpha1.ObjectSync) {
 
 func sendEvents(err error, nodeName string, sync *v1alpha1.ObjectSync, resourceType string,
 	objectResourceVersion string, obj interface{}) {
+	// some workaround to keep the CRD resource type consistent with the edge side
+	// TODO: investigate to have a more systematic way to handle.
```
Comment (pdgetrf) [inline cloud/pkg/synccontroller/objectsync.go:97]: let's create an issue for todo. TODOs in code tend to get ignored and lost.

File: common/constants/default.go [hunk 5]
```diff
@@ -137,7 +144,24 @@ const (
 	CSIOperationTypeControllerUnpublishVolume = "controllerunpublishvolume"
 	CSISyncMsgRespTimeout                     = 1 * time.Minute
 
+	ResourceTypeMission     = "mission"
+	ResourceTypeMissionList = "missionlist"
+
+	ResourceTypeEdgeCluster     = "edgecluster"
+	ResourceTypeEdgeClusterList = "edgelcjusterlist"
```
Comment (pdgetrf) [inline common/constants/default.go:151]: edgelcjusterlist --> edgeclusterlist

File: edge/pkg/clusterd/edgecluster_status_reporter.go [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+
+	ec.Name = config.Config.Name
+	ec.Spec.Kubeconfig = config.Config.Kubeconfig
+	ec.Spec.KubeDistro = config.Config.KubeDistro
+
+	ec.Labels = map[string]string{
+		// Kubernetes built-in labels
+		v1.LabelHostname: ec.Name,
+
+		// KubeEdge specific labels
```
Comment (pdgetrf) [inline edge/pkg/clusterd/edgecluster_status_reporter.go:80]: KubeEdge --> Fornax

File: edge/pkg/clusterd/helper/helper.go [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+	testClusterCommand := fmt.Sprintf("%s cluster-info --kubeconfig=%s", config.Config.KubectlCli, config.Config.Kubeconfig)
+	if _, err := util.ExecCommandLine(testClusterCommand); err != nil {
+		klog.Errorf("The cluster is unreachable: %v", err)
+		return false
+	}
+
+	return true
+}
+
+// this is a hacky way for PoC only
```
Comment (pdgetrf) [inline edge/pkg/clusterd/helper/helper.go:79]: does this comment still apply? 

if yes, we need to create an issue to track this

File: edge/pkg/clusterd/helper/helper.go [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+	}
+	if b == nil && len(a) == 0 {
+		return true
+	}
+
+	// for other cases, use the regular map compare
+	return reflect.DeepEqual(a, b)
+}
+
+// for some reason we still need to find out, the same mission spec objects may be no longer deep-equal.
```
Comment (pdgetrf) [inline edge/pkg/clusterd/helper/helper.go:129]: please create an issue for this. very intriguing to find out.

File: edge/pkg/clusterd/mission_deployer.go [hunk 9]
```diff
@@ -0,0 +1,348 @@
...
+	mission, err := helper.GetMissionByName(name)
+	if err != nil {
+		return err
+	}
+
+	return m.DeleteMission(mission)
+}
+
+func (m *MissionDeployer) isMatchingMission(mission *edgeclustersv1.Mission) bool {
+	// if the placement field is empty, it matches all the edge clusters
```
Comment (pdgetrf) [inline edge/pkg/clusterd/mission_deployer.go:130]: discussion: a wild card should be used to match all the edge cluster, and an an empty field should match none.

also, in future, access control should be used so a user wouldn't be able to deploy to "all" the clusters, e.g. those that the user doesn't own. 

let's have a separate design on placement someday.

File: edge/pkg/clusterd/mission_deployer.go [hunk 10]
```diff
@@ -0,0 +1,348 @@
...
+
+	return fmt.Errorf("Hit the errors in mission align: %v", errs)
+}
+
+// create a yaml to use by "kubectl apply" command
+func buildMissionYaml(input *edgeclustersv1.Mission) (string, error) {
+	// probably due to the json encoder in arktos, the commmnd "kubectl apply missiong" in arktos
+	// fails if the mission.StateCheck.Command is nil or empty.
+	// We trick it with a string with one space.
+	// TODO: We need a non-hacky way.
```
Comment (pdgetrf) [inline edge/pkg/clusterd/mission_deployer.go:189]: create a new issue for this.

File: edge/pkg/metamanager/process.go [hunk 11]
```diff
@@ -109,7 +130,19 @@ func resourceUnchanged(resType string, resKey string, content []byte) bool {
 }
 
 func (m *metaManager) processInsert(message model.Message) {
+	resKey, resType, _ := parseResource(message.GetResource())
```
Comment (pdgetrf) [inline edge/pkg/metamanager/process.go:133]: what does "res" stand for in these variables? probably "response"? 

maybe just use key and type as name instead. "res" doesn't add too much values to the names.

File: hack/local-up-kubeedge.sh [hunk 12]
```diff
@@ -36,6 +36,12 @@ function check_prerequisites {
 function kind_up_cluster {
   echo "Running kind: [kind create cluster ${CLUSTER_CONTEXT}]"
   kind create cluster ${CLUSTER_CONTEXT}
+
+  # re-install flannel as we run into kubeletNotReady issues from time to time due to NetPlugin not ready
+  sudo rm -rf /opt/cni/bin/
+  sudo rm -rf /etc/cni/net.d/
+  kubectl apply -f https://github.com/coreos/flannel/raw/master/Documentation/kube-flannel.yml
+  sleep 10
```
Comment (pdgetrf) [inline hack/local-up-kubeedge.sh:44]: this is quite odd that we need to do this. nice job figuring it out.

**Review by pdgetrf — APPROVED** [review]
looks good overall. 

a few questions in place.

**Comment by pdgetrf** [comment]
ma this is a long PR. needed two coffee breaks to finish it. haha.

this is a great start for the repo! 

the ppt docs you did for this PR and design, let's also check them into this repo (later is okay).

File: cloud/pkg/edgecontroller/controller/downstream.go (hunk 1 above)
Comment (chenqianfzh) [reply cloud/pkg/edgecontroller/controller/downstream.go:107]: updated.

File: cloud/pkg/edgecontroller/controller/upstream.go (hunk 2 above)
Comment (chenqianfzh) [reply cloud/pkg/edgecontroller/controller/upstream.go:563]: deleted.

File: cloud/pkg/missionstatepruner/pruner.go (hunk 3 above)
Comment (chenqianfzh) [reply cloud/pkg/missionstatepruner/pruner.go:124]: Changed it to "Updating"

File: common/constants/default.go (hunk 5 above)
Comment (chenqianfzh) [reply common/constants/default.go:151]: fixed.

File: edge/pkg/metamanager/process.go (hunk 11 above)
Comment (chenqianfzh) [reply edge/pkg/metamanager/process.go:133]: actually, "res" stands for "resource" here. This short form has been used throughout this file . Let's just keep it as is here for code consistency.

File: edge/pkg/clusterd/mission_deployer.go (hunk 10 above)
Comment (chenqianfzh) [reply edge/pkg/clusterd/mission_deployer.go:189]: After a second thought, I believe the current trick is acceptable. I removed this "TODO" comment.

File: edge/pkg/clusterd/mission_deployer.go (hunk 9 above)
Comment (chenqianfzh) [reply edge/pkg/clusterd/mission_deployer.go:130]: Yep, we need a separate design on this. 

Currently, we are re-using the constraint expression in Kubernentes federation, which may not be applicable to the situation of edge clsuters.

issue created to track this: https://github.com/CentaurusInfra/fornax/issues/26

File: cloud/pkg/synccontroller/objectsync.go (hunk 4 above)
Comment (chenqianfzh) [reply cloud/pkg/synccontroller/objectsync.go:97]: tracking issue : https://github.com/CentaurusInfra/fornax/issues/27

File: cloud/pkg/synccontroller/objectsync.go [hunk 13]
```diff
@@ -26,14 +27,35 @@ import (
 func (sctl *SyncController) manageObject(sync *v1alpha1.ObjectSync) {
 	var object metav1.Object
 
-	gv, err := schema.ParseGroupVersion(sync.Spec.ObjectAPIVersion)
-	if err != nil {
-		return
+	var gvr schema.GroupVersionResource
+	var resourceNamespace string
+	// A workaround to get the GVR for the CRD types in edgeclusters.kubeedge.io group
+	// A better solution is to register the CRD types with the scheme
```
[the remark this answers is not in the document]
Comment (chenqianfzh) [reply cloud/pkg/synccontroller/objectsync.go:33]: created a tracking issue: https://github.com/CentaurusInfra/fornax/issues/27. Will remove this comment when this issue is fixed.

File: edge/pkg/clusterd/edgecluster_status_reporter.go (hunk 6 above)
Comment (chenqianfzh) [reply edge/pkg/clusterd/edgecluster_status_reporter.go:80]: Actually, they are Kubeedge specific. These labels are used in edge nodes as well.

File: edge/pkg/clusterd/helper/helper.go (hunk 7 above)
Comment (chenqianfzh) [reply edge/pkg/clusterd/helper/helper.go:79]: Created the tracking issue: https://github.com/CentaurusInfra/fornax/issues/28

File: edge/pkg/clusterd/helper/helper.go (hunk 8 above)
Comment (chenqianfzh) [reply edge/pkg/clusterd/helper/helper.go:129]: Now I know this issue is due to Json encoder. Json encoding a nil array into an empty array, thus the deep-equal conditions no long hold. 

I rewrote the comment to explain this.

**Pull request merged**